initial ieee802.15.4 API

japaric opened the pull request:
what the title says

this PR adds an `ieee802154::Radio` API that wraps the `RADIO` peripheral and puts it in IEEE 802.15.4 mode.

I haven't checked the silicon erratas of (or tested) other chips (the code contains one workaround for (nRF52840 specific?) silicon errata) so for now I have put the API directly under `nrf52840-hal`. But I guess the API should also work on other nRF52 devices -- I think the nRF51 chips do not have an 802.15.4 mode? And I don't know about other chip families.

The implementation is not super IEEE compliant. It's missing a backoff delay mechanism when doing CCA (Clear Channel Assessment) before transmitting packets and it's also missing IFS (inter-frame spacing) between consecutive packets (the hardware has some built-in support for this but the PR doesn't expose it). But it is sufficient to send packets around and receive them on other nRF52s running the same implementation, though it would be good to check this against some third party implementation like the DWM1001 or the MRF24J40.

There are some TODOs here and there in terms of additional things the API could expose like changing the transmit power. I'm not sure if all those need to be added before a review though. They could be added as follow up PRs too.

TODO items
- [x] API to change TX power (the default value doesn't go very far, like tens of centimer or so in a noisy 2.4 GHz environment)
- [x] API to change the start of frame delimiter (the default value matches the IEEE spec)
- [x] API to extract the LQI (Link Quality Indicator)
- [x] verify that CRC are being checked. May need to send a malformed packet from a different device (MRF24J40?)
- [x] "The total usable payload (PSDU) is 127 octets, but when CRC is being used, this is reduced to 125 octets of usable payload." -- adjust MAX_LEN accordingly

thalesfragoso commented:
Thanks for this work.

> But I guess the API should also work on other nRF52 devices -- I think the nRF51 chips do not have an 802.15.4 mode? And I don't know about other chip families.

Both the nRF52810 and nRF52832 don't support 802.15.4 mode. 51 probably doesn't support it either.

Yatekii commented:
Cool stuff! I need to test this on my boards =)

The 52840, 52811 52833 and 5340 support 802.15.4.

Are you aware of the https://crates.io/crates/ieee802154 crate? I think it is higher level (MAC) layer, but could be useful to test this and not reinvent the wheel.

Thanks for putting in the work!

japaric commented:
> Are you aware of the https://crates.io/crates/ieee802154 crate? I think it is higher level (MAC) layer, but could be useful to test this and not reinvent the wheel.

That's the MAC frame. Yes, that's what you should put in the payload of the `Packet` defined here (which is actually the PHY layer frame) if you want to be IEEE compliant.

Also, yes that would be useful to test this against the MRF24J40, which (iirc) does do HW-level MAC address filtering. Though that may end up being more of a test for the `ieee802154` crate itself.

hannobraun commented:
@japaric 
> though it would be good to check this against some third party implementation like the DWM1001 or the MRF24J40

The DW1000 uses UWB (ultra-wideband) in the PHY layer, so I don't think it can communicate with the nRF radio (I might be wrong wrong; despite my work with the DW1000/DWM1001, I know very little about radios). 

> Also, yes that would be useful to test this against the MRF24J40, which (iirc) does do HW-level MAC address filtering. Though that may end up being more of a test for the `ieee802154` crate itself.

The DW1000 also does hardware-based address matching and this is used in the driver. To my knowledge, this has never caused any problems with `ieee802154`.

jonathanpallant commented:
Yes, the DW1000 is a UWB PHY (so around 500 MHz bandwidth at 3 to 10 GHz depending on channel) and hence incompatible with a 2.4 GHz PHY (around 80 MHz bandwidth at 2.4 GHz) based system.

I wasn't aware the DW1000 had hardware address detection, so that's good to know!

Getting back on topic, Rust 802.15.4 on the nRF52 series is excellent. Great work :)

blueluna commented:
Nice! Looks better than my old implementation.
https://github.com/blueluna/nrf52840-dk-experiments/tree/master/nrf52-radio-802154
I hope to try it out when I get time.

japaric commented:
for those not getting notifications on new commits being pushed: I have added APIs to change the TX power and the start of frame delimiter; added a method the get a packet's Link Quality Indicator; and added CRC verification to the Radio.recv method (the method now returns a Result)

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+/// IEEE 802.15.4 channels
+///
+/// NOTE these are NOT the same as WiFi 2.4 GHz channels
+// TODO it is possible to use non-standard frequencies below 2_400 MHz; should those be exposed too?
```
This list is correct. To be compliant with IEEE 802.15.4 other frequencies shall not be exposed. The hardware in selected nRF devices is capable of using other frequencies, but it should not be used in 802.15.4 protocol.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+
+            radio.radio.pcnf1.write(|w| {
+                w.maxlen()
+                    .bits(Packet::MAX_LEN + 2 /* CRC */) // payload length
```
This line is a little confusing. When we configure hardware, we usually work on PHY layer. Max PSDU length is 127 and it includes CRC. I would change this constant to `MAX_PSDU_LEN = 127` and use it here without `+ 2`. It would be unambiguous.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+    pub fn set_channel(&mut self, channel: Channel) {
+        self.disable();
+
+        // NOTE(unsafe) radio is currently disabled
```
It is OK to change frequency while radio is enabled. The FREQUENCY register is double buffered and it is read on RXEN or TXEN tasks.
There is no need to call `self.disable();` before.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+
+    /// Changes the Clear Channel Assessment method
+    pub fn set_cca(&mut self, cca: Cca) {
+        self.disable();
```
Like with frequency it is OK to change CCA mode while radio is enabled.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+
+    /// Changes the Start of Frame Delimiter
+    pub fn set_sfd(&mut self, sfd: u8) {
+        // FIXME don't completely turn off the radio; RXIDLE or TXIDLE are probably OK
```
I expect that every state is OK, I think this register is also double buffered, but I'm not sure about it, because we have never modified SFD value.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+    }
+
+    /// Moves the radio from any state to the RXEN state
+    fn enable_rx(&mut self) {
```
There is a possibility that with RADIO SHORTS enabled this function would work in an unexpected way. I would add at least assertion here to verify that selected shorts are disabled.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+
+        // start transfer
+        dma_start_fence();
+        self.radio.tasks_start.write(|w| w.tasks_start().set_bit());
```
That's exacly the place where I would clear `Event::End` flag, to be sure that it is not unintentionally set by some previously executed operation.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+        self.radio.tasks_start.write(|w| w.tasks_start().set_bit());
+
+        // wait until we have received something
+        self.wait_for_event(Event::End);
```
I'm not sure what is the design of this driver and whether it is OK, but please note that this is blocking and it can be blocked for a very long time (infinite in theory, if this is the only device in the network).

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+                .bit_is_set()
+            {
+                // channel is busy
+                self.radio.events_ccaidle.reset();
```
Typo: `events_ccaidle` -> `events_ccabusy`

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+    /// This method performs Clear Channel Assessment (CCA) first and sends the `packet` only if the
+    /// channel is observed to be *clear* (no transmission is currently ongoing), otherwise no
+    /// packet is transmitted and the `Err` variant is returned
+    pub fn try_send(&mut self, packet: &Packet) -> Result<(), ()> {
```
It is good to use `CCAIDLE_TXEN` short for this operation to make sure that the transmission takes place right after CCA. With the design of this function, there is unknown software delay between CCAIDLE event and TXEN task.

hubertmis replied:
I see it is applied in `send`. Is there any reason here it is handled by software?

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+
+    /// Tries to send the given `packet`
+    ///
+    /// This method performs Clear Channel Assessment (CCA) first and sends the `packet` only if the
```
Please note, that this method shall not be used for ACK TX, because you shall not perform CCA when transmitting ACKs

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+            loop {
+                if self
+                    .radio
+                    .events_ccaidle
```
You could wait for PHYEND instead. There is no need to wait here for CCAIDLE and for PHYEND later

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+                match state {
+                    STATE_A::DISABLED => return,
+
+                    STATE_A::RXRU | STATE_A::RXIDLE | STATE_A::TXRU | STATE_A::TXIDLE => {
```
`RXIDLE` state may indicate ongoing CCA procedure that may result in TX with radio shorts. To be fully safe, following set of tasks should be triggered:
`CCASTOP`
`STOP`
`DISABLE`

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+    }
+
+    /// Moves the radio from any state to the DISABLED state
+    fn disable(&mut self) {
```
There is a possibility that with RADIO SHORTS enabled this function would work in an unexpected way. I would add at least assertion here to verify that selected shorts are disabled.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+
+/// An IEEE 802.15.4 packet
+///
+/// This `Packet` is closest to the PPDU (PHY Protocol Data Unit) defined in the IEEE spec. The API
```
`closest to the PPDU` is ambiguous. I would state here clearly that this packet is PHR concatenated with PSDU

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+    ///
+    /// NOTE `src` data will be truncated to `MAX_PACKET_SIZE` bytes
+    pub fn copy_from_slice(&mut self, src: &[u8]) {
+        let len = cmp::min(src.len(), Self::MAX_LEN as usize) as u8;
```
I don't think it is a good idea to silently limit the given slice size. When user of this function wants to transmit 200 bytes packet and calls this function, would get only 125 bytes transmitted and would not be notified that there was a problem. I would change function signature to return `Result` or add assertion.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,630 @@
+    ///
+    /// NOTE `len` will be truncated to `MAX_LEN` bytes
+    pub fn set_len(&mut self, len: u8) {
+        let len = cmp::min(len, Self::MAX_LEN);
```
Again silently truncated. Using such API could be a source of hard to find bugs in the higher layers.

japaric commented:
@hubertmis thanks for the great feedback! (I have seen it just now). I'll incorporate it in the PR in the coming days.

hubertmis commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,662 @@
+    /// Maximum amount of usable payload (CRC excluded) a single packet can contain, in bytes
+    pub const CAPACITY: u8 = 125;
+    const CRC: u8 = 2; // size of the CRC, which is *never* copied to / from RAM
+    const MAX_PSDU_LEN: u8 = Self::CAPACITY + 2 /* Self::CRC */;
```
```suggestion
    const MAX_PSDU_LEN: u8 = Self::CAPACITY + Self::CRC;
```
I'm not rust expert yet and I'm not sure if this syntax would cause any problems, but it looks much simpler to replace a magic number and a comment with the constant mentioned by the comment.

japaric replied:
ah, yes. I meant to use the constant here. Thanks for catching this.

hubertmis commented:
@japaric , to verify CRC of generated frames, you can easily use https://github.com/NordicSemiconductor/nRF-Sniffer-for-802.15.4 . Firmware used in this project reports only frames with valid CRC.

hubertmis commented:
I was thinking about timing out while receiving. Current implementation blocks execution of the calling thread until a frame is received or a timeout fires. It's not good for the system performance.

In drivers in C usually asynchronous approach is implemented: function like `recv` changes the driver state to 'receive' and stops execution. If the driver receives a frame, it calls a callback asynchronously. But it looks that this asynchronous callback approach is not perfect fit for Rust, because I think it does not go well with futures or programs using async/await.

Recently I've spotted https://github.com/rust-embedded/nb and it looks like a solution to this problem. The user of the driver could call `recv()` just to enter receive state. Later the user could call a new function like `read()` that would implement this nb scheme and return a frame if it was received before the function was called, or `WouldBlock` if the RX buffer is empty. With such API, the calling thread would not be blocked while waiting for frames and futures, async/await, or other models could be used.

What do you think about it?

jonas-schievink commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,804 @@
+        }
+    }
+
+    /// Recevies one radio packet and copies its contents into the given `packet` buffer
```
```suggestion
    /// Receives one radio packet and copies its contents into the given `packet` buffer
```

jonas-schievink commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,804 @@
+    }
+
+    /// Sample the received signal power (i.e. the presence of possibly interfering signals)
+    /// within the bandwidth of the currently used channel for sample_cycles iterations.
```
```suggestion
    /// within the bandwidth of the currently used channel for `sample_cycles` iterations.
```

jonas-schievink commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,804 @@
+                self.radio.events_edend.reset();
+
+                // note that since we have increased EDCNT, the EDSAMPLE register contains the
+                // maximumrecorded value, not the average
```
```suggestion
                // maximum recorded value, not the average
```

jonas-schievink commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,804 @@
+    /// Returns an empty packet (length = 0)
+    // XXX I believe that be making this *not* `const` it is not possible to place a `Packet` in
+    // `.rodata` (modulo unsafe `#[link_section]` shenanigans) thus it is not necessary to check the
+    // address of packet in `Radio.{send,recv}` (EasyDMA can only operate on RAM addresses)
```
Hmm, I think I'd prefer not to rely on this. Instead, `recv` and `send` could both take a `&mut Packet`, which ensures that it's in writeable RAM (other modules in the HAL already skip the address check when they deal with mutable data).

jonas-schievink commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,804 @@
+pub const DEFAULT_CCA: Cca = Cca::CarrierSense;
+
+/// Default radio channel = Channel 20 (`2_450` MHz)
+pub const DEFAULT_CHANNEL: Channel = Channel::_18; // _17 todo undo; set this to interfere with my wifi
```
It looks like this needs to be reset to channel 17.

Are the defaults specified somewhere or were they picked arbitrarily?

japaric replied:
the FREQUENCY register reset value is `2`, which doesn't match to any of the `Channel` values because they are all multiples of 5.
The channel values (the frequencies) are the standard IEEE frequencies.

japaric replied:
changed the default channel to 11 (2405 MHz) which is closer to the reset value of the FREQUENCY register (2402 MHz)

jonas-schievink commented on nrf52840-hal/src/ieee802154.rs:
```diff
@@ -0,0 +1,804 @@
+    // RADIO needs to be (re-)enabled to pick up new settings
+    needs_enable: bool,
+    // used to freeze `Clocks`
+    _clocks: &'c (),
```
This could be `PhantomData` to shrink `Radio` by a pointer.

jonas-schievink commented:
Okay, I think this is ready to go. The remaining notes can be addressed later, and an async/non-blocking API can also be added then.

bors r+

jonas-schievink commented:
I only just noticed that the module was added directly to the nrf52840-hal instead of nrf-hal-common. I've opened https://github.com/nrf-rs/nrf-hal/pull/299 to move it (and enable it for the nRF52833 in the process).
